c/parallel/src/transform.cu

```diff
@@ -0,0 +1,617 @@
+
+CUresult cccl_device_unary_transform_build(
+  cccl_device_transform_build_result_t* build_ptr,
+  cccl_iterator_t input_it,
+  cccl_iterator_t output_it,
+  cccl_op_t op,
+  int cc_major,
+  int cc_minor,
+  const char* cub_path,
+  const char* thrust_path,
+  const char* libcudacxx_path,
+  const char* ctk_path)
+{
+  CUresult error = CUDA_SUCCESS;
+
+  try
+  {
+    const char* name = "test";
+
+    const int cc                 = cc_major * 10 + cc_minor;
+    const auto policy            = transform::get_policy();
+    const auto input_it_value_t  = cccl_type_enum_to_name<input_storage_t>(input_it.value_type.type);
+    const auto output_it_value_t = cccl_type_enum_to_name<output_storage_t>(output_it.value_type.type);
+    const auto offset_t          = cccl_type_enum_to_name(cccl_type_enum::CCCL_INT64);
+    const std::string input_iterator_src =
+      make_kernel_input_iterator(offset_t, "input_iterator_t", input_it_value_t, input_it);
+    const std::string output_iterator_src =
+      make_kernel_output_iterator(offset_t, "output_iterator_t", output_it_value_t, output_it);
+    const std::string op_src = make_kernel_user_unary_operator(input_it_value_t, output_it_value_t, op);
+
+    constexpr std::string_view src_template = R"XXX(
+#define _CUB_HAS_TRANSFORM_UBLKCP 0
+#include <cub/device/dispatch/kernels/transform.cuh>
+struct __align__({1}) input_storage_t {{
+  char data[{0}];
+}};
+struct __align__({2}) output_storage_t {{
+  char data[{3}];
+}};
+{8}
+{9}
+struct prefetch_policy_t {{
+  static constexpr int block_threads = {4};
+  static constexpr int items_per_thread_no_input = {5};
+  static constexpr int min_items_per_thread      = {6};
+  static constexpr int max_items_per_thread      = {7};
+}};
+struct device_transform_policy {{
+  struct ActivePolicy {{
+    static constexpr auto algorithm = cub::detail::transform::Algorithm::prefetch;
+    using algo_policy = prefetch_policy_t;
+  }};
+}};
+{10}
+)XXX";
+
+    const std::string& src = std::format(
+      src_template,
+      input_it.value_type.size, // 0
+      input_it.value_type.alignment, // 1
+      output_it.value_type.size, // 2
+      output_it.value_type.alignment, // 3
+      policy.block_threads, // 4
+      policy.items_per_thread_no_input, // 5
+      policy.min_items_per_thread, // 6
+      policy.max_items_per_thread, // 7
+      input_iterator_src, // 8
+      output_iterator_src, // 9
+      op_src); // 10
+
+#if false // CCCL_DEBUGGING_SWITCH
+    fflush(stderr);
+    printf("\nCODE4NVRTC BEGIN\n%sCODE4NVRTC END\n", src.c_str());
+    fflush(stdout);
+#endif
+
+    std::string kernel_name = transform::get_kernel_name(input_it, output_it, op);
+    std::string kernel_lowered_name;
+
+    const std::string arch = std::format("-arch=sm_{0}{1}", cc_major, cc_minor);
+
+    constexpr size_t num_args  = 8;
+    const char* args[num_args] = {
+      arch.c_str(), cub_path, thrust_path, libcudacxx_path, ctk_path, "-rdc=true", "-dlto", "-default-device"};
```

Comment:
Out of curiosity, why do we need `-default-device`? We don't use it in any of the other `c.parallel` algorithms. Maybe add a comment explaining why?

Reply:
I found it was needed because of the use of unqualified lambdas in `transform.cuh`:

https://github.com/NVIDIA/cccl/blob/a74629ec718621b69bbb1629da930019732b3fec/cub/cub/device/dispatch/kernels/transform.cuh#L102

Now, if I were to qualify these lambdas with `__device__`, then `nvcc` asks me to pass the `--expt-extended-lambda` flag - which works. The problem is, that flag isn't supported by NVRTC as far as I could tell.

I'll add a comment explaining this.